R/aggregate_game_stats_def.R

```diff
@@ -0,0 +1,527 @@
+#' `qb_epa`, the function `add_qb_epa()` will be called to add it.
+#' @param weekly If `TRUE`, returns week-by-week stats, otherwise, stats
+#' for the entire Data frame.
+#' @description Build columns that aggregate official passing, rushing, and receiving stats
+#' either at the game level or at the level of the entire data frame passed.
+#' @return A data frame including the following columns (all ID columns are
+#' decoded to the gsis ID format):
+#' \describe{
+#' \item{player_id}{ID of the player. Use this to join to other sources.}
+#' \item{player_name}{Name of the player}
+#' \item{games}{The number of games where the player recorded passing, rushing or receiving stats.}
+#' \item{recent_team}{Most recent team player appears in `pbp` with.}
+#' \item{season}{Season if `weekly` is `TRUE`}
+#' \item{week}{Week if `weekly` is `TRUE`}
+#' }
+#' @export
+#' @seealso The function [load_player_stats()] and the corresponding examples
+#' on [the nflfastR website](https://www.nflfastr.com/articles/nflfastR.html#example-11-replicating-official-stats)
+#' @examples
+#' \donttest{
+#'   pbp <- nflfastR::load_pbp(2020)
+#'
+#'   weekly <- calculate_player_stats_def(pbp, weekly = TRUE)
+#'   dplyr::glimpse(weekly)
+#'
+#'   overall <- calculate_player_stats_def(pbp, weekly = FALSE)
+#'   dplyr::glimpse(overall)
+#' }
+#'
+
+#++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+# what do we need:
+#
+# Solo Tackles --> done
+# Tackles With Assist --> done
+# Assisted Tackles --> done
+# Tackles for Loss --> done
+# TFL Yards --> done
+# Sacks --> done
+# Sack Yards --> done
+# QB Hits --> done
+# Passes Defensed --> done
+# Interceptions --> done
+# Interception Yards --> done
+# Interception Return TDs ///// --> only "TD" for defense
+# Forced Fumbles --> done
+# Opp Fumble Recoveries --> done
+# Opp Fumble Recovery Yards --> done
+# Opp Fumble Recovery TDs ///// --> only "TD" for defense
+# Safeties --> done
+# Penalties --> done
+# Penalty Yards /////
+# Fumbles --> done
+# Own Fumble Recoveries --> done
+# Own Fumble Recovery Yards --> done
+# Own Fumble Recovery TDs ///// --> only "TD" for defense
+
+calculate_player_stats_def <- function(pbp, weekly = FALSE) {
+
+  # need newer version of nflreadr to use load_players
+  rlang::check_installed("nflreadr (>= 1.3.0)")
+
+  # Prepare data ------------------------------------------------------------
+
+  suppressMessages({
+    # 1. for "normal" plays: get plays that count in official stats
+    # we exclude special teams and 2pts here for now
+    data <- pbp %>%
+      dplyr::filter(
+        !is.na(.data$down),
+        .data$play_type %in% c("pass", "qb_kneel", "qb_spike", "run")
+      ) %>%
+      nflfastR::decode_player_ids()
+
+    # 2. filter penalty plays that are no_plays for penalty stats
+    penalty_data <- pbp %>%
+      dplyr::filter(
+        !is.na(.data$down),
```

Comment:
We need to remove lines 87 and 88 here. 
Line 87 removes penalties on kickoffs and extra points (we don't want that).
Line 88 removes penalties accepted on top of the play (unnecessary roughness, etc.). We don't want that either.

Not sure if this causes issues downstream, but we need to loosen this filter to just` .data$penalty == 1`

Reply:
Done in upcoming commit